doc/decisions/header_file_structure.md

```diff
@@ -12,31 +12,95 @@ The big problem is kdbprivate.h. It has two main problems:
 
 ## Constraints
 
+- Each library should have at least one separate header file:
+
+  Without this constraint, we would constantly include things that aren't needed.
+  This is inconvenient as it clutters the auto-suggestions of IDEs.
+  It also slows down the compile process, since the compiler has to parse everything that was included (see [also](https://lore.kernel.org/lkml/YdIfz+LMewetSaEB@gmail.com/T/)).
+
+- Non-public API and public API should be kept in separate files
+- Some non-public APIs need to be accessible for testing, such headers should not be packaged/installed
+
 ## Assumptions
 
+- There are different categories of "private":
```

Comment:
This is a very important point, actually it is a decision on its own.

Reply:
What exactly would the decision part of the decision be? I'm just stating facts here. The assumption is that all symbols can be put into one of these four groups. But I'm not sure there is anything to decide here. The decision to use these four groups to build our headers is what **this** decision is about. Not sure, what you want to separate here.

Reply:
It is probably more standard to use the word "protected" for something that is "public" within Elektra's repository but private for the outside world (point 3).

> The assumption is that all symbols can be put into one of these four groups.

Then write what you mean:

```suggestion
- We assume that all parts of the API belong to one of the following visibility categories:
```

> Not sure, what you want to separate here.

The separate decision would be about which parts belong to which "visibility category".  You give good examples here.

Reply:
Hmm honestly I don't think we should do a seperate decision here. The examples that @kodebach listed serve illustrative purposes in this decision. 

Creating decisions just for documenting where which each function of each library should be placed is too much IMHO. That sounds very much like documentation for documentation's sake. Checking the placement is something that should be part of the PR review.

Reply:
If we apply the suggestion by @markus2330 then we must also define 'public' and 'internal' visibilities in this section.